### PR TITLE
Update the grammar to support match

### PR DESCRIPTION
**Pull request by ewang12**
This adds pattern matching to the grammar, but doesn't implement anything yet.

**Status:** In development

**Fixes:** N/A

### CODE REVIEW

File: src/parser.cpp [hunk 1]
```diff
@@ -22,14 +22,24 @@
...
+    Let = Pattern EQUALS Term SEPARATOR Term
     DataType = DATA LEFT_PAREN DataConstructorList RIGHT_PAREN
     DataConstructorList = | DataConstructor DataConstructorTail
     DataConstructorTail = | SEPARATOR DataConstructor DataConstructorTail
     DataConstructor = IDENTIFIER DataConstructorParams
     DataConstructorParams = | IDENTIFIER DataConstructorParams
     Member = Term DOT IDENTIFIER
     Group = LEFT_PAREN Term RIGHT_PAREN
+    Pattern =
+      VariablePattern | ConstructorPattern | LEFT_PAREN Pattern RIGHT_PAREN
```
Comment (ewang12) [inline src/parser.cpp]: Right now the grammar permits extra parentheses at the top level. For instance,
```
(x) = something
```
We can remove `LEFT_PAREN Pattern RIGHT_PAREN` if we want to disallow this.

Comment (stepchowfun) [inline src/parser.cpp]: I vote for keeping this.

File: src/parser.cpp [hunk 2]
```diff
@@ -22,14 +22,24 @@
     Variable = IDENTIFIER
     Abstraction = IDENTIFIER ARROW Term
     Application = Term Term
-    Let = IDENTIFIER EQUALS Term SEPARATOR Term
+    Let = Pattern EQUALS Term SEPARATOR Term
```
Comment (stepchowfun) [inline src/parser.cpp]: If we are changing `Let`, we might as well change `Abstraction` too, right?

(But for the MVP, I wouldn't change either.)

File: src/parser.cpp [hunk 3]
```diff
@@ -22,14 +22,24 @@
...
     DataConstructorParams = | IDENTIFIER DataConstructorParams
     Member = Term DOT IDENTIFIER
     Group = LEFT_PAREN Term RIGHT_PAREN
+    Pattern =
+      VariablePattern | ConstructorPattern | LEFT_PAREN Pattern RIGHT_PAREN
+    VariablePattern = IDENTIFIER
+    ConstructorPattern = Term DOT IDENTIFIER PatternList
+    PatternList = PatternInList | PatternInList PatternList
+    PatternInList = VariablePattern | LEFT_PAREN ConstructorPattern RIGHT_PAREN
+    Match = MATCH Term LEFT_PAREN CaseList RIGHT_PAREN
```
Comment (stepchowfun) [inline src/parser.cpp]: We should add `Match` as one of the cases for `Term`.

File: src/parser.cpp [hunk 4]
```diff
@@ -22,14 +22,24 @@
...
     DataConstructor = IDENTIFIER DataConstructorParams
     DataConstructorParams = | IDENTIFIER DataConstructorParams
     Member = Term DOT IDENTIFIER
     Group = LEFT_PAREN Term RIGHT_PAREN
+    Pattern =
+      VariablePattern | ConstructorPattern | LEFT_PAREN Pattern RIGHT_PAREN
+    VariablePattern = IDENTIFIER
+    ConstructorPattern = Term DOT IDENTIFIER PatternList
+    PatternList = PatternInList | PatternInList PatternList
+    PatternInList = VariablePattern | LEFT_PAREN ConstructorPattern RIGHT_PAREN
```
Comment (stepchowfun) [inline src/parser.cpp]: Why do we need `PatternInList`? Why not just use `Pattern`?

File: src/parser.cpp (hunk 3 above)
Comment (ewang12) [inline src/parser.cpp]: Good catch.

File: src/parser.cpp (hunk 4 above)
Comment (ewang12) [inline src/parser.cpp]: There was ambiguity with nested patterns. For example, if you have a tree data type and your pattern is
```
tree.node tree.node left right value tree.empty
```
How would this be parsed? The parser can't know how to group the patterns, so `PatternInList` requires that `ConstructorPattern`s be surrounded with parentheses.

Come to think of it though, we want to be able to have patterns like
```
tree.node tree.empty value tree.empty
```
without parentheses. So I think I actually need to add another case to the grammar.

Comment (stepchowfun) [inline src/parser.cpp]: Oh ok. I thought this was supposed to be the "raw" unambiguous grammar, and the one with ambiguities resolved is below.

**Comment by stepchowfun** [comment]
Grammar looks good! Let's implement the parser.

**Pull request closed without merging**